### DISCUSSION TITLE
What's the actual price of the AI service?

### DISCUSSION THREAD

**Discussion by osseonews** [nhost/nhost#2628] [Q&A]
I'm trying to understand what the pricing is for AI on your Pro plan. You say it's based on compute time, but then it says $50/Month? So is this $50/Month minimum no matter what, even if we don't use it, and then it's extra beyond the flat $50 for actual usage?

**Comment by dbarrosop** [comment]
Current pricing is based on resources allocated. For instance, if you allocate 0.125vCPUs and 256MB RAM it should be ~6.25$ month. Other than that you need your own OpenAI key and they will charge you based on your usage of their chatgpt and other services.

**Reply by osseonews** [reply]
OK, so why does it say $50/Month on the page in big font? If it's billed per minute why is the charge $50? This is confusing and misleading. I am concerned as I have got heavily burned on other Saas platforms that claim to charge per usage (e.g. Hasura Cloud), but actually charge you even if you never use the service at all. Just turning it on counts as usage even if you don't make any calls. The per usage fee is on top of some base fee. Also, how would I know what to allocate? Or how to estimate any potential costs?

**Reply by dbarrosop** [reply]
where does it say 50$/month? We charge for resources allocated, we don't charge for usage. If you allocate 1vCPU, we charge for 1vCPU (50$ month) regardless if you make 1 request or 1 billion requests. You are free to allocate more or less resources at any time based on your needs as resources are billed per minute intervals. For instance:

<img width="972" alt="Screenshot 2024-04-02 at 11 14 58" src="https://github.com/nhost/nhost/assets/6246622/3de30430-30c8-4ec0-a71c-ceb5e035bbf4">

It says approximate cost because some months have more minutes than others and because you can change them anytime so the price isn't really set in stone.

P.S: For more details on pricing check https://nhost.io/pricing

**Reply by osseonews** [reply]
Where does it say $50/Month? Right on the website.

![pricing-nhost](https://github.com/nhost/nhost/assets/26583659/ea21aaed-e4bf-4038-96fa-0a243abdbdaf)

**Reply by dbarrosop** [reply]
Per vCPU / core / month. You can allocate partial cores which brings the cost down proportionally. Easiest is to use the estimator in settings -> AI

**Reply by osseonews** [reply]
Your pricing is very confusing and misleading. You are not charging per minute because there is no way an actual user is going to be turning this feature on and off during a month. There are approximately 43,800 minutes in a month, so at .0012 charge per minute for 2GB, the actual bill for anyone using this service will be $52.56 per month, even if they run very few requests. I'm assuming you need at least 2GB of memory for run any normal application or why would it say that? Honestly, I don't know because it's not clear. Would this feature work with 256 memory? All in all, this seems like an amazing service but i'd be very hesitant to even turn it on because the billing isn't clear at all. This totally reminds me of Hasura Cloud. they claim to price at $1.50 per "active hour". What this actual means, though is that if you just set up the service and never make one API call, you will be charged around $1,000 for the month because nobody actually sets up the service and then spends their time turning it on and off during the month. Same applies here. you should just say it's $50 per month for unlimited calls (2gb memory, which I assume is the minium to actually have this function in a normal production) and not make it sound like you are paying per minute.

**Reply by dbarrosop** [reply]
> You are not charging per minute because

We are charging per minute because if you run this for 31 minutes that's how much you will be charged. Also because if you allocate 1 core for 29 minutes and half a core for the rest of the month, then we charge 1 core for 29 minutes and half a core for the rest of the month.

> There are approximately 43,800 minutes in a month

Depends on the month, in any case, we say approximately for a reason.

> or why would it say that?

Why would it say what exactly? I am not sure I am following your line of thought here.

> Would this feature work with 256 memory?

It depends on your usage, we certainly have users using it with only that amount of memory. Our settings page wouldn't suggest that starting point if it wasn't enough to run the service.

> This totally reminds me of Hasura Cloud. they claim to price at ...

I can't comment on other people's pricing models

> you should just say it's $50 per month for unlimited calls

That wouldn't be correct though as you are making a lot of assumptions that aren't correct.

**Reply by osseonews** [reply]
"We are charging per minute because if you run this for 31 minutes that's how much you will be charged."
In what reasonable scenario would anyone turn this on for 31 minutes. I guess if someone is just a developing something as a hobby, they will turn this on and off all the time. But, that's not how it will work for real use cases. 

Anyway, I'll just assume I'll paying $50 per month if I turn this on, as that seems to be the minimum fee we would incur under a realistic scenario.

**Reply by osseonews** [reply]
Sorry one other quick question. It seems like you are using OpenAi for autoembeddings. this can become super expensive and I don't think it's necessary for the vectors at all. What other autoembeddings are you working on?

**Reply by dbarrosop** [reply]
> In what reasonable scenario would anyone turn this on for 31 minutes

Dev, staging, QA, I tried it and didn't like it... We charge per minute so you know you are not stuck with your choice. This is more relevant for production when you have a flash sale or some other event and need more resources temporarily (i.e. you double your resources during black Friday or similar event).

> Anyway, I'll just assume I'll paying $50 per month

My recommendation would be to start with fewer resources. I doubt you need a full core + 2GB of RAM for this service. In any case, you can start with whatever resources you feel comfortable with and then you can use the observability dashboards to see CPU/RAM usage and adapt accordingly.

> It seems like you are using OpenAi for autoembeddings. this can become super expensive and I don't think it's necessary for the vectors at all

How much data do you have? Curious because in our experience, generating the embeddings for 30.000 rows was <1$. From their pricing:

- ada v2 - $0.10 / 1M tokens

Which isn't too bad but I guess it all depends on the amount of data you have to generated embeddings for.

> What other autoembeddings are you working on?

We are not currently working with any alternative service, if you have any in mind feel free to propose. You can also leave the "query" and "mutation" fields empty when configuring auto-embeddings and then we won't generate them for you, we will only mange the search and similarity queries. Not as convenient but it gives you an option if you want to generate them with your own infrastructure, for instance.

**Reply by osseonews** [reply]
Thanks. Yeah, I just looked at OpenAi and didn't realize they had a new embedding, text-embedding-3-small, which is very cheap. we'll use that.

Another follow up question, usually for example you need to combine a title and content for the embedding, see the OpenAi example, https://cookbook.openai.com/examples/get_embeddings_from_dataset ("We will combine the review summary and review text into a single combined text. The model will encode this combined text and it will output a single vector embedding.").

How is this handled in Nhost, do we create a rule that the auto embedding is done by combining certain fields in the table?

**Reply by dbarrosop** [reply]
You will need to provide a graphql query we use to extract the data from the database. Then we use that data to generate the embeddings. In the end, we will send a json object to openai like:

```
{
   "title": "Some title",
   "contents: "this is the content..."
}
```

Assuming you provided a graphql query with the fields `title` and `contents`. You can find more details here:

https://docs.nhost.io/guides/ai/auto-embeddings
